diff --git a/target/linux/ramips/image/mt7620.mk b/target/linux/ramips/image/mt7620.mk
index 9f3a8dce0e..7913da62b0 100644
--- a/target/linux/ramips/image/mt7620.mk
+++ b/target/linux/ramips/image/mt7620.mk
@@ -1423,6 +1423,27 @@ define Device/youku_x2
 endef
 TARGET_DEVICES += youku_x2

+define Device/xiaomi_miwifi-r3
+  SOC := mt7620a
+  BLOCKSIZE := 128k
+  PAGESIZE := 2048
+  KERNEL_SIZE := 4096k
+  IMAGE_SIZE := 32768k
+  UBINIZE_OPTS := -E 5
+  IMAGES += kernel1.bin rootfs0.bin breed-factory.bin factory.bin
+  IMAGE/kernel1.bin := append-kernel | check-size $$$$(KERNEL_SIZE)
+  IMAGE/rootfs0.bin := append-ubi | check-size
+  IMAGE/sysupgrade.bin := sysupgrade-tar | append-metadata
+  IMAGE/factory.bin := append-kernel | pad-to $$(KERNEL_SIZE) | append-ubi | check-size
+  IMAGE/breed-factory.bin := append-kernel | pad-to $$(KERNEL_SIZE) | \
+                            append-kernel | pad-to $$(KERNEL_SIZE) | \
+                            append-ubi | check-size
+  DEVICE_VENDOR := Xiaomi
+  DEVICE_MODEL := Mi Router R3
+  DEVICE_PACKAGES := kmod-mt76x2 kmod-usb2 kmod-usb-ohci uboot-envtools
+endef
+TARGET_DEVICES += xiaomi_miwifi-r3
+
 define Device/youku_yk-l1
   SOC := mt7620a
   IMAGE_SIZE := 32448k
diff --git a/target/linux/ramips/mt7620/base-files/etc/board.d/02_network b/target/linux/ramips/mt7620/base-files/etc/board.d/02_network
index 577f77cb93..53f2db5b74 100644
--- a/target/linux/ramips/mt7620/base-files/etc/board.d/02_network
+++ b/target/linux/ramips/mt7620/base-files/etc/board.d/02_network
@@ -1,3 +1,4 @@
+9272

 . /lib/functions.sh
 . /lib/functions/uci-defaults.sh
@@ -202,6 +203,10 @@ ramips_setup_interfaces()
                ucidef_add_switch "switch0" \
                        "0:lan:2" "1:lan:1" "4:wan" "6@eth0"
                ;;
+       xiaomi,miwifi-r3)
+               ucidef_add_switch "switch0" \
+                       "1:lan" "4:lan" "0:wan" "6@eth0"
+               ::
        lenovo,newifi-y1s)
                ucidef_add_switch "switch0" \
                        "1:lan:4" "2:lan:3" "4:lan:2" "5:lan:1" "0:wan" "6@eth0"
@@ -425,6 +430,10 @@ ramips_setup_macs()
                wan_mac=$(mtd_get_mac_binary Factory 0x2e)
                label_mac=$wan_mac
                ;;
+       xiaomi,miwifi-r3)
+               wan_mac=$(mtd_get_mac_binary factory 0x28)
+               lan_mac=$(macaddr_setbit_la "$wan_mac")
+               ;;
        tplink,archer-c2-v1|\
        tplink,archer-c20-v1|\
        tplink,archer-c20i|\
diff --git a/target/linux/ramips/mt7620/base-files/lib/upgrade/platform.sh b/target/linux/ramips/mt7620/base-files/lib/upgrade/platform.sh
index 6dd7fc7cef..d0a92f952d 100755
--- a/target/linux/ramips/mt7620/base-files/lib/upgrade/platform.sh
+++ b/target/linux/ramips/mt7620/base-files/lib/upgrade/platform.sh
@@ -38,6 +38,11 @@ platform_do_upgrade() {
                                of=/dev/mtdblock$idx
                default_do_upgrade "$1"
                ;;
+       xiaomi,miwifi-r3)
+               # this make it compatible with breed
+               dd if=/dev/mtd0 bs=64 count=1 2>/dev/null | grep -qi breed && CI_KERNPART_EXT="kernel_stock"
+               nand_do_upgrade "$1"
+               ;;
        *)
                default_do_upgrade "$1"
                ;;
diff --git a/target/linux/ramips/mt7620/target.mk b/target/linux/ramips/mt7620/target.mk
index 34488f95af..1d1d281a0f 100644
--- a/target/linux/ramips/mt7620/target.mk
+++ b/target/linux/ramips/mt7620/target.mk
@@ -4,7 +4,7 @@

 SUBTARGET:=mt7620
 BOARDNAME:=MT7620 based boards
-FEATURES+=usb ramdisk
+FEATURES+=usb nand ramdisk
 CPU_TYPE:=24kc

 DEFAULT_PACKAGES += kmod-rt2800-soc wpad-basic-mbedtls swconfig
